From c6c19d6cd9c708a76f0074115ccb5482a4f7190a Mon Sep 17 00:00:00 2001
From: Robert Kanter <rkanter@apache.org>
Date: Wed, 27 Jan 2016 17:11:07 -0800
Subject: [PATCH 1139/2748] MAPREDUCE-6431. JobClient should be an
 AutoClosable (haibochen via rkanter)

(cherry picked from commit dca0dc8ac28e843acd8b79c9560245638a539fde)
(cherry picked from commit a99fa744909ef69b968f24def5138d2f3beb568f)

Change-Id: I6ba3979496014bc8238918e45696b8737dff4166
---
 .../org/apache/hadoop/mapred/TestJobClient.java    |   10 ++++++++++
 .../java/org/apache/hadoop/mapred/JobClient.java   |    3 ++-
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-common/src/test/java/org/apache/hadoop/mapred/TestJobClient.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-common/src/test/java/org/apache/hadoop/mapred/TestJobClient.java
index b18b531..bf37b03 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-common/src/test/java/org/apache/hadoop/mapred/TestJobClient.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-common/src/test/java/org/apache/hadoop/mapred/TestJobClient.java
@@ -87,4 +87,14 @@ public void testGetStagingAreaDir() throws IOException, InterruptedException {
         client.getClusterHandle().getStagingAreaDir().toString()
             .equals(client.getStagingAreaDir().toString()));
   }
+
+  /**
+   * Asks the compiler to check if JobClient is AutoClosable.
+   */
+  @Test(timeout = 10000)
+  public void testAutoClosable() throws IOException {
+    Configuration conf = new Configuration();
+    try (JobClient jobClient = new JobClient(conf)) {
+    }
+  }
 }
diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java
index 84edfeb..34b1293 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapred/JobClient.java
@@ -137,7 +137,7 @@
  */
 @InterfaceAudience.Public
 @InterfaceStability.Stable
-public class JobClient extends CLI {
+public class JobClient extends CLI implements AutoCloseable {
 
   @InterfaceAudience.Private
   public static final String MAPREDUCE_CLIENT_RETRY_POLICY_ENABLED_KEY =
@@ -488,6 +488,7 @@ public JobClient(InetSocketAddress jobTrackAddr,
   /**
    * Close the <code>JobClient</code>.
    */
+  @Override
   public synchronized void close() throws IOException {
     cluster.close();
   }
-- 
1.7.9.5

